File: gui/src/types/goal.ts
import type { Origin } from "@/types/conversation";

/**
 * Goal v2 state machine (.scratch/goal-simplify/PRD.md §3.2), mirroring
 * Rust `api::GoalStatus`.
 *
 *   - `active`   — Core re-dispatches a continuation on every idle.
 *   - `paused`   — user aborted the run, or Core restarted. Recoverable:
 *                  the session's next user message resumes it.
 *   - `blocked`  — the model declared a blocker, or the run errored.
 *                  Recoverable the same way; the difference is who judged.
 *   - `completed` / `budget_limited` / `stopped` / `failed` — terminal.
 */
export type GoalStatus =
  | "active"
  | "paused"
  | "blocked"
  | "completed"
  | "budget_limited"
  | "stopped"
  | "failed";

/** Open = still owns the session's idle time, or can get it back with
 * one user message. The set the per-session uniqueness index guards. */
export const OPEN_GOAL_STATUSES: readonly GoalStatus[] = [
  "active",
  "paused",
  "blocked",
];

export function isOpenGoalStatus(status: GoalStatus): boolean {
  return OPEN_GOAL_STATUSES.includes(status);
}

export function isTerminalGoalStatus(status: GoalStatus): boolean {
  return !isOpenGoalStatus(status);
}

/** Wire twin of Rust `api::GoalBrief` (camelCase JSON). */
export interface GoalBrief {
  id: string;
  /** The session this goal drives — exactly one, for its whole life. */
  sessionId: string;
  objective: string;
  status: GoalStatus;
  /** Time ceiling in seconds. Absent = no ceiling. */
  budgetSeconds?: number;
  startedAt: string;
  endedAt?: string;
  pausedAt?: string;
  latestSummary?: string;
  resultSeenAt?: string;
  /** Continuations Core has dispatched so far (the wrap-up one included). */
  continuationCount: number;
  /** True once the budget-limit wrap-up continuation went out. */
  wrapUpDispatched: boolean;
  /** Wall-clock seconds from `startedAt` to `endedAt` (terminal) or to
   * the read (open). Computed by Core, never stored — the GUI reads it
   * straight instead of running its own clock. */
  elapsedSeconds: number;
  createdAt: string;
  updatedAt: string;
  /** Who set the goal. Absent for GUI-set goals. */
  origin?: Origin;
}

/** Input for the `start_session_goal` command. `budgetSeconds: null` is
 * the explicit "no ceiling" choice; the GUI always sends the field. */
export interface StartSessionGoalInput {
  sessionId: string;
  objective: string;
  budgetSeconds?: number | null;
}

/** Persisted message row echoed back by `start_session_goal`. The GUI
 * does NOT render it from here — the same row arrives through
 * `user-message-persisted`, which is the single mirror path. */
export interface GoalObjectiveMessage {
  id: string;
  sessionId: string;
  role: "user" | "agent" | "system";
  content: string;
  createdAt: string;
  summary?: string;
  turnIndex?: number;
  goalId?: string;
  origin?: Origin;
}

export interface StartSessionGoalResult {
  goal: GoalBrief;
  message: GoalObjectiveMessage;
  /** Always `"dispatched"` — a dispatch failure is an error, never a
   * half-started goal. */
  dispatch: string;
}

/** What the confirm dialog resolves to. `budgetSeconds: null` = no
 * ceiling (an explicit user choice, never a default). */
export interface GoalLaunchConfig {
  budgetSeconds: number | null;
}

File: core/src/api/goal.rs
use serde::{Deserialize, Serialize};

use super::{Origin, SessionId};

/// Default time ceiling when a caller asks for "the default" (the GUI's
/// recommended preset, the CLI's `--budget-minutes` fallback). `None` on
/// [`CreateGoalInput::budget_seconds`] means no ceiling at all — that is
/// an explicit choice, never a default.
pub const DEFAULT_GOAL_BUDGET_SECONDS: u32 = 60 * 60;

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct GoalId(pub String);

impl GoalId {
    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl std::fmt::Display for GoalId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

/// Goal v2 state machine (.scratch/goal-simplify/PRD.md §3.2).
///
/// - `Active`: Core re-dispatches a continuation whenever the session goes
///   idle.
/// - `Paused`: the user aborted the current run, or Core restarted with the
///   goal active. Recoverable — the next user message on the session
///   resumes it.
/// - `Blocked`: the model declared a blocker, or the run ended in an error.
///   Recoverable the same way as `Paused`; the distinction is who judged.
/// - `Completed` / `BudgetLimited` / `Stopped` / `Failed`: terminal.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GoalStatus {
    Active,
    Paused,
    Blocked,
    Completed,
    BudgetLimited,
    Stopped,
    Failed,
}

impl GoalStatus {
    /// Not terminal: the goal still owns its session's idle time (`Active`)
    /// or can get it back with one user message (`Paused` / `Blocked`).
    /// This is the set the per-session uniqueness index guards.
    pub fn is_open(self) -> bool {
        matches!(self, Self::Active | Self::Paused | Self::Blocked)
    }

    pub fn is_terminal(self) -> bool {
        !self.is_open()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GoalBrief {
    pub id: GoalId,
    /// The session this goal drives. A goal has exactly one; the session
    /// may carry many goals over its life, at most one of them open.
    pub session_id: SessionId,
    pub objective: String,
    pub status: GoalStatus,
    /// Time ceiling in seconds. `None` = no ceiling.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub budget_seconds: Option<u32>,
    pub started_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ended_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub paused_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub latest_summary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result_seen_at: Option<String>,
    /// Continuations Core has dispatched so far (the wrap-up one included).
    pub continuation_count: u32,
    /// True once the budget-limit wrap-up continuation went out; the next
    /// idle after it lands the goal in `BudgetLimited`.
    pub wrap_up_dispatched: bool,
    /// Wall-clock seconds from `started_at` to `ended_at` (terminal) or to
    /// now (open). Computed at read time, never stored. Paused time is not
    /// subtracted in v2.
    pub elapsed_seconds: u64,
    pub created_at: String,
    pub updated_at: String,
    /// Who set the goal. Absent for GUI-set goals (same convention as
    /// `SessionBrief.origin`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub origin: Option<Origin>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateGoalInput {
    pub session_id: SessionId,
    pub objective: String,
    /// `None` = no ceiling. Callers that want the product default pass
    /// [`DEFAULT_GOAL_BUDGET_SECONDS`] explicitly.
    #[serde(default)]
    pub budget_seconds: Option<u32>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn open_and_terminal_partition_the_status_set() {
        for s in [GoalStatus::Active, GoalStatus::Paused, GoalStatus::Blocked] {
            assert!(s.is_open());
            assert!(!s.is_terminal());
        }
        for s in [
            GoalStatus::Completed,
            GoalStatus::BudgetLimited,
            GoalStatus::Stopped,
            GoalStatus::Failed,
        ] {
            assert!(s.is_terminal());
            assert!(!s.is_open());
        }
    }

    #[test]
    fn status_serializes_snake_case() {
        assert_eq!(
            serde_json::to_string(&GoalStatus::BudgetLimited).unwrap(),
            "\"budget_limited\""
        );
    }
}
